### DISCUSSION TITLE
Compress, Decompress pickle file

### DISCUSSION THREAD

**Discussion by krisvive** [pycaret/pycaret#1116] [Q&A] [answered]
Hi All,

My final pickle file (which includes entire transformation pipeline + trained model) is too big in size, ~15GB. To compress and then share this file, I tried 2 approaches...
```
import joblib
joblib.dump(rf,  'finalized_model.pkl.gz',compress=3)
```
and...
```
import gzip, pickle
with gzip.open('finalized_model.pkl.gz', 'wb') as ofp:
    pickle.dump(rf, ofp)
```
I was able to successfully compress with both methods, file was approx. 500 MB
However, when I loaded the file successfully, with either of the 2 methods...
`uncompressed_model = joblib.load(r'C:\Users\Administrator\RMF_2.0_VK\Models\rf_AllErrors_16032021.pkl.gz')`
or
```
with gzip.open('rf_AllErrors_16032021.pkl.gz', 'rb') as ifp:
    uncompressed_model = pickle.load(ifp)
```
and check the contents of this uncompressed file, they are as below...
```
print(uncompressed_model)
print(type(uncompressed_model))
```
Outcome...
**_RandomForestClassifier(n_jobs=-1, random_state=1246)
<class 'sklearn.ensemble._forest.RandomForestClassifier'>_**
When I am trying to load this uncompressed file and to make predictions...
`predictions = predict_model(uncompressed_model, data = unseen_data)`
I am getting below error...
**_NameError: name 'prep_pipe' is not defined
ValueError: Pipeline not found_**

Question: Can anyone please guide on the best approach to compress and decompress pycaret (2.3.0) pickle file (transformation pipeline + trained model)? Looking at the above message, it seems only the model is getting compressed and decompressed later, and transformation pipeline is getting missed. Appreciate some help on this.

**Comment by Yard1** [comment, marked as answer]
We can add a way to pass parameters to save_model that will let you specify the compression setting.

**Reply by krisvive** [reply]
that would be helpful. is there a parameter already available in save_model()? or planned for future release ?

**Reply by Yard1** [reply]
You can install the master branch to get access to it right away: `pip install -U git+https://github.com/pycaret/pycaret.git`

**Reply by krisvive** [reply]
Thanks @Yard1 
So I need to update my save_model() as below? Trying to run the entire pipeline and then finally below command...

`save_model(final_model, 'pickle_filename',**{"value":final_model, "filename":'pickle_filename.pkl.gz',"compress":4})`

**Reply by Yard1** [reply]
`save_model(final_model, 'pickle_filename', compress=4)` This is all that is necessary.

**Reply by krisvive** [reply]
cool, thanks @Yard1

**Reply by krisvive** [reply]
Hi @Yard1 I installed the latest version from master branch using the command you have above and restarted my kernel. It shows version as pycaret==2.3.0 and now when using compress parameter under save model, am getting below error...
![image](https://user-images.githubusercontent.com/6905855/112593690-dd282400-8e2d-11eb-83b0-0e21203fc050.png)

Very much appreciate your support

**Reply by Yard1** [reply]
Is it possible for you to restore the environment to the default settings and reinstall then?

**Reply by krisvive** [reply]
> Is it possible for you to restore the environment to the default settings and reinstall then?

Sure, let me try doing that

**Reply by krisvive** [reply]
Hi @Yard1 - Instead of this compress-decompress approach to shrink my model size and transfer it...I tried other approach of just saving only the model (with _model_only=True_ parameter in save_model()) but even with it the pickle file size is similar to the model saved without this parameter. Since I am saving only the model and not transformation pipeline, I was expecting a smaller pickle file size.

Also, when I checked the datatype, contents of both model only pickle file and default pickle file (using commands like below); both seems to be same.
```
saved_model_mo = load_model("Model_Only_PickleFile")
print(saved_model_mo)
type(saved_model_mo)
```
Both the file's datatype is sklearn.pipeline.Pipeline

Any help here, much appreciated ?

**Reply by Yard1** [reply]
The model would be far bigger than the rest of the pipeline, so that is to be expected. Have you managed to get compress-decompress running? If not, could you create a code snippet that can reproduce the problem? I can verify it then.

**Reply by krisvive** [reply]
got it, thanks
I reinstalled pycaret too and tried. getting below error. Hope this snipped will help you to understand better and try to resolve issue...
![image](https://user-images.githubusercontent.com/6905855/112626570-63f0f700-8e56-11eb-98a1-09ece8ea3ca9.png)
P.S: just ignore the difference in pickle file name, i have replaced it

**Reply by Yard1** [reply]
Thanks, will take a look and get back to you

**Reply by Yard1** [reply]
I couldn't reproduce the issue with PyCaret installed with git. Please check this notebook: https://colab.research.google.com/drive/1oK2ApHXSQMSfn_Qzpf2iTFaahbBX6Pqd?usp=sharing

**Reply by krisvive** [reply]
> I couldn't reproduce the issue with PyCaret installed with git. Please check this notebook: https://colab.research.google.com/drive/1oK2ApHXSQMSfn_Qzpf2iTFaahbBX6Pqd?usp=sharing

Thanks @Yard1 Very Helpful ! Let me try again

**Reply by krisvive** [reply]
It worked fine when I uninstalled PyCaret 2.3 and installed again using the above link you shared.
